### PR TITLE
DRS localizer thin slice [BT-192]

### PR DESCRIPTION
**Pull request by mcovarr**
Some refactoring: 

* Separated the concepts of resolving and downloading.
* `DrsLocalizerMain#resolve` now chooses the `Downloader` implementation based on the content of the Martha response.
* The two downloader implementations support access and GCS URLs respectively.
* Tests for the downloader implementations have been separated from tests for the resolver.

### CODE REVIEW

**Review by kshakir — COMMENTED** [review]
For "DRAFT" PR debugging purposes, perhaps temporarily change the martha_v3 project to `broad-dsde-cromwell-dev`?
 https://github.com/broadinstitute/cromwell/blob/60/src/ci/resources/papi_application.inc.conf.ctmpl#L74

File: src/ci/resources/papi_v2_hg_reference_images.inc.conf [hunk 1]
```diff
@@ -0,0 +1,102 @@
+reference-disk-localization-manifests = [
```
Comment (breilly2) [inline src/ci/resources/papi_v2_hg_reference_images.inc.conf:1]: This looks unrelated.

Comment (mcovarr) [reply src/ci/resources/papi_v2_hg_reference_images.inc.conf:1]: oops

File: cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala [hunk 2]
```diff
@@ -127,32 +94,37 @@ class MockDrsLocalizerMain(drsUrl: String,
...
+      case MockDrsPaths.fakeDrsUrlWithoutAnyResolution =>
+        marthaResponse
+      case MockDrsPaths.fakeDrsUrlWithAccessUrlResolutionOnly =>
+        marthaResponse.copy(
+          accessUrl = Option(AccessUrl(url = "http://abc/def/ghi.bam", headers = None)))
+      case MockDrsPaths.fakeDrsUrlWithAccessUrlAndGcsResolution =>
+        marthaResponse.copy(
+          accessUrl = Option(AccessUrl(url = "http://abc/def/ghi.bam", headers = None)),
+          gsUri = Option("gs://some/uri"))
+      case _ => throw new RuntimeException("fudge")
```
Comment (mcovarr) [inline cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala:127]: um

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala [hunk 3]
```diff
@@ -37,114 +33,38 @@ class DrsLocalizerMain(drsUrl: String,
...
-  private final val RequesterPaysErrorMsg = "Bucket is requester pays bucket but no user project provided."
-  private final val ExtractGcsUrlErrorMsg = "No resolved url starting with 'gs://' found from Martha response!"
+  private final val ExtractUriErrorMsg = "No access URL nor GCS URI starting with 'gs://' found in Martha response!"
 
-  def getGcsDrsPathResolver: IO[GcsLocalizerDrsPathResolver] = {
+  def getDrsPathResolver: IO[LocalizerDrsPathResolver] = {
     IO {
       val drsConfig = DrsConfig.fromEnv(sys.env)
-      new GcsLocalizerDrsPathResolver(drsConfig)
+      new LocalizerDrsPathResolver(drsConfig)
```
Comment (kshakir) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:41]: Very very ToL: While renaming, should `LocalizerDrsPathResolver` be called `DrsLocalizerDrsPathResolver` to match `DrsLocalizerMain`?

See also: http://java.metagno.me/

Or maybe `DrsLocalizerMain` just becomes something like `Main`. 🤷

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import com.typesafe.scalalogging.StrictLogging
+
+import scala.sys.process.{Process, ProcessLogger}
+
+case class AccessUrlDownloader(accessUrl: AccessUrl, downloadLoc: String) extends Downloader with StrictLogging {
+
+  def generateDownloadScript(): String = {
+    val signedUrl = accessUrl.url
+    // TODO headers, requester pays, refinements to retry strategy for certain 4xx statuses
+    s"""mkdir -p $$(dirname '$downloadLoc') && curl --location --retry 3 --retry-connrefused --retry-delay 10 --fail --output '$downloadLoc' '$signedUrl'"""
```
Comment (kshakir) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:14]: ToL: The change for requester pays will require changes to the request sent to martha, not changes to this line here. But ToL I think it's fine to lump the TODOs together.

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+  override def download: IO[ExitCode] = {
+    val signedUrl = accessUrl.url
+      // TODO probably don't want to log the actual signed URL
+      logger.info(s"Attempting to download $signedUrl to $downloadLoc")
```
Comment (kshakir) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:20]: When is this TODO going to be addressed? Before the PR merges or does it need a future Jira ticket?

ToL: One can discuss with appsec as they might have recommendations of the safety of options like:
- log the URL anyway since it expires, or...
- log the URL without the query string

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GcsUriDownloader.scala [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      requesterPaysProjectIdOption match {
+        case Some(userProject) =>
+          s"""if [ "$$RC_GSUTIL" != "0" ]; then
+             |  # Check if error is requester pays. If yes, retry gsutil copy using project flag
+             |  if grep -q '$RequesterPaysErrorMsg' gsutil_output.txt; then
+             |    echo "Received 'Bucket is requester pays' error. Attempting again using Requester Pays billing project"
+             |    ${gcsCopyCommand(s"-u $userProject")} > gsutil_output.txt 2>&1
+             |    RC_GSUTIL=$$?
+             |  fi
+             |fi""".stripMargin
```
Comment (kshakir) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GcsUriDownloader.scala:75]: 1 of 2: Any opinions on this block terminating with a newline? (the change would affect the test expectations too.)
```scala
|fi
|""".stripMargin
```

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GcsUriDownloader.scala [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+       |
+       |${recoverWithRequesterPays()}
+       |
+       |if [ "$$RC_GSUTIL" != "0" ]; then
+       |  echo "Failed to download the file. Error: $$(cat gsutil_output.txt)" >&2
+       |  exit "$$RC_GSUTIL"
+       |else
+       |  echo "Download complete!"
+       |  exit 0
+       |fi""".stripMargin
```
Comment (kshakir) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GcsUriDownloader.scala:98]: 2 of 2: Any opinions on this block terminating with a newline? (the change would affect the test expectations too.)
```scala
|fi
|""".stripMargin
```

This one terminates the actual full bash script. When I'm not trying to code golf I generally [prefer newlines](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file#answer-18789).

**Review by kshakir — APPROVED** [review]

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GcsUriDownloader.scala (hunk 7 above)
Comment (mcovarr) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/GcsUriDownloader.scala:98]: ⛳

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala (hunk 5 above)
Comment (breilly2) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:20]: We touched on this during tech-talk this morning. Since this is in the localizer, I'm not sure where these logs are going... maybe an object in the workspace bucket, maybe Google Cloud Logging... either way, it probably doesn't matter. Access to these objects are tied to an individual's connection to the data, which is even more specific than workspace access. Therefore, I think we should not log the full signed URL.

I think we should still log _something_ to aid debugging of localization. Logging the DRS URI and the decision to download via signed URL might be enough... or maybe chopping off any query string from the signed URL would make it safe enough to log.


... though I just realized that this is the _localizer_ we're talking about. Where do these logs go?

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala [hunk 8]
```diff
@@ -37,114 +33,38 @@ class DrsLocalizerMain(drsUrl: String,
...
-
-  def resolveAndDownload(): IO[ExitCode] = {
-    val fields = NonEmptyList.of(MarthaField.GsUri, MarthaField.GoogleServiceAccount)
+  def resolve(): IO[Downloader] = {
+    val fields = NonEmptyList.of(MarthaField.GsUri, MarthaField.GoogleServiceAccount, MarthaField.AccessUrl)
     for {
-      localizerGcsDrsPathResolver <- getGcsDrsPathResolver
-      marthaResponse <- localizerGcsDrsPathResolver.resolveDrsThroughMartha(drsUrl, fields)
-      // Currently Martha only supports resolving DRS paths to GCS paths
-      gcsUrl <- IO.fromEither(marthaResponse.gsUri.toRight(new RuntimeException(ExtractGcsUrlErrorMsg)))
```
Comment (breilly2) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:145]: Did we lose this check along the way? The code is gone, but maybe it's rephrased somewhere and I'm not recognizing it.

**Review by breilly2 — COMMENTED** [review]
Looking good. A few comments/questions before I approve.

File: cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala (hunk 8 above)
Comment (mcovarr) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala:145]: This condition now maps to [this line](https://github.com/broadinstitute/cromwell/blob/bt_186_bdc_public_drs_localizer_thin_slice/cromwell-drs-localizer/src/main/scala/drs/localizer/DrsLocalizerMain.scala#L65) where we have neither an access URL nor a GS URI.

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala (hunk 5 above)
Comment (mcovarr) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:20]: I believe this output should make it to the `task_name.log` file that winds up under the user's "folder" for the associated task in GCS ([example log](https://00f74ba44b93a7a7fc4c2de08ea29c0b680795f665-apidata.googleusercontent.com/download/storage/v1/b/cloud-cromwell-dev-self-cleaning/o/cromwell_execution%2Fci%2Fcall_cache_cha_cha%2F06e44cfb-8d6f-47ab-a9b0-57f902eb0ae3%2Fcall-read_files_new_command%2Fread_files_new_command.log?jk=AFshE3W5xy0F1pE5Dfo2_7MuIr4UTVn5KmtNhjrimzxI47kL7hDvdV7FRYCN7WvfLnTvxjblyyXEW5l-GyvwPT27UvxKrD-z1poemc-UlncMM4wAh17iPrWY6kg_IH3m5oVjkhvDFBwWBGp33qdb_PpiyAe_jeJnvnLfe9poX-IXBUgbj2EML_RZ3BEcxxavAb3BwikM8lgYTrmAJEitqed-WyfEkhQfGYU3OQZ2mSpFjkolin4d6Xzue_47lsYP5kY_UilDBAvpKASqMXO0YFakLhXb2so5NsxvejzrtPuTS6bZPPgfz5--e0yZ0MyvDdbaYm-oy1YySUUrgiPRyB6sy8CeQGXq5BKHYi1Tt1uBvIAhVB0z2LL3sAn4cik7UmcP2OHeEQwyGZPjtcYCxtltda4g_yLcSMHQuTZPszVHeATJo19nHWf4rJ0nCAnv8rgdUMLb-E9lnUbMHqK2G33B9Qk3TsEggR5nw21ICBSAYYTl5eu1a63iBSeyenzlyMoRHhftB-RsAWk0Kwi6dE1VNYdwtErRqRYoFq6u5pg7uXiCuXBiB7ZkLrQ131w_5Fo2akwKZmHNWlDoV5NxBaIF-wPWWSyQ9ilOxgz-Z7RrNTPlcK2LtVoEgWT5FE_074xrs_4w9VZ-TXi3kO_GLHkl3GVDl41YRbCR78W-l9EPwpYRwQ_Jdl3qMqmi5iC3iRjmmIcUQjyp75lHxIiu86uzUBEgKV1FQGNBDTOFVxlS0zO0LA7cZXco54n0C6mMmkYmIHns13kZplfHqEbvPMUnWdR8DSkWB7nldlR7WAWIJLXdb7kiu82ShhM2rrxsT9FfoaVfoo8x6wxO42RB-7M-sigSuTNjNadiJ8AhS9_1FsNdWdHHDqK_D3GCo3dFkAO8cP0LMMEQb7iFvDUzxffrgeAcEsKsy-OK8zsqZqkn73n7X39yTLcXSbyn0QdqJBeihrm4B0Hq6ffP-l25hfJmRfbrpT4fgKs2NLTYhTx8hYRTXticsE6yttdBIWYmoTXkJIUFlyYb1eSqHqMtWZjbNXMJvz_XRVXDyOXieW6oG_r2vCsUofmjixEdCHr83pDmVnTqDqWeY3mKrcwpwdhaJIldfAPxBYnd0M3UPmshN3-pYjbT-gbpVA8iHCuEXa-t0Eda79E&isca=1) from a CI run). This should be a very safe place as it will already contain other sensitive data such as the outputs from this task. So I'm kinda wondering if we really need to change this line after all...

Comment (kshakir) [reply cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:20]: The logs for the DRS localizers (AccessURL or GCS) go into the same GCS bucket that contains the other call logs. Below are steps to get to the GCS DRS localizer log from the CI for this current PR.

- Start with the [Travis Logs](https://travis-ci.com/github/broadinstitute/cromwell/jobs/499764041)
- Search for `Submitting drs_usa_jdr returned workflow id`
- Get workflow ID = `a2699805-0a16-479f-bacb-f2bfa8755a4b`
- CI uses at output [root](https://github.com/broadinstitute/cromwell/blob/61/src/ci/resources/papi_v1_v2alpha1_provider_config.inc.conf#L2) of `gs://cloud-cromwell-dev-self-cleaning/cromwell_execution/ci`
- So path to this workflow outputs is `gs://cloud-cromwell-dev-self-cleaning/cromwell_execution/ci/drs_usa_jdr/a2699805-0a16-479f-bacb-f2bfa8755a4b`
- Path to aggregated Papi action logs for the call is `gs://cloud-cromwell-dev-self-cleaning/cromwell_execution/ci/drs_usa_jdr/a2699805-0a16-479f-bacb-f2bfa8755a4b/call-skip_localize_jdr_drs_with_usa/skip_localize_jdr_drs_with_usa.log`
- The individual download action just happened to be action 13 (one can iterate through the actions until one finds the individual action logs, or can use the gcloud papi operations info to list the actions) revealing the individual action log is `gs://cloud-cromwell-dev-self-cleaning/cromwell_execution/ci/drs_usa_jdr/a2699805-0a16-479f-bacb-f2bfa8755a4b/call-skip_localize_jdr_drs_with_usa/pipelines-logs/action/13/stdout`

File: cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala [hunk 9]
```diff
@@ -15,109 +14,77 @@ class DrsLocalizerMainSpec extends AnyFlatSpec with CromwellTimeoutSpec with Mat
...
+    val mockDrsLocalizer = new MockDrsLocalizerMain(MockDrsPaths.fakeDrsUrlWithoutAnyResolution, fakeDownloadLocation, None)
 
     the[RuntimeException] thrownBy {
-      mockDrsLocalizer.resolveAndDownload().unsafeRunSync()
-    } should have message "No resolved url starting with 'gs://' found from Martha response!"
+      mockDrsLocalizer.resolve().unsafeRunSync()
+    } should have message "No access URL nor GCS URI starting with 'gs://' found in Martha response!"
+  }
+
+  it should "resolve to use the correct downloader for a gs uri" in {
```
[the remark this answers is not in the document]
Comment (mcovarr) [reply cromwell-drs-localizer/src/test/scala/drs/localizer/DrsLocalizerMainSpec.scala:55]: yup sure does

File: cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  def generateDownloadScript(): String = {
+    val signedUrl = accessUrl.url
+    // TODO headers, refinements to retry strategy for certain 4xx statuses
+    s"""mkdir -p $$(dirname '$downloadLoc') && curl --location --retry 3 --retry-connrefused --retry-delay 10 --fail --output '$downloadLoc' '$signedUrl'"""
+  }
+
+  override def download: IO[ExitCode] = {
+    val signedUrl = accessUrl.url
+      // We don't want to log the actual signed URL here. On a PAPI backend this log will end up under the user's
+      // workspace bucket, but that bucket may have visibility different than the data referenced by the signed URL.
```
Comment (mcovarr) [inline cromwell-drs-localizer/src/main/scala/drs/localizer/downloaders/AccessUrlDownloader.scala:20]: Upon further reflection I removed the signed URL from this log statement completely.

**Review by breilly2 — APPROVED** [review]
👍   Looks good to me!

**Review by kshakir — APPROVED** [review]
Re 👍

**Comment by mcovarr** [comment]
Merge to develop gated on a BT-219 release of Martha that supports access urls.
